Add a new package for IcedTea

snehring opened the pull request:
This is based heavily on #2624 and adds the alsa-lib and cups dependencies, if you'd like those separated into their own pull requests let me know.

This requires an existing icedtea or openjdk environment to compile. There's supposedly a way to accomplish this using ecj, but given that that's not included with gcc anymore it didn't seem like the best road to go down.

adamjstewart commented on var/spack/repos/builtin/packages/icedtea/package.py:
```diff
@@ -0,0 +1,163 @@
+    variant('X', default=False, description="Build with GUI support.")
+    variant('shenandoah', default=False, description="Build with the shenandoah gc. Only for version 3+")
+
+    depends_on('pkg-config', type='build')
```
If I recall, it needed `pkg-config@0.9.0:`

adamjstewart commented on var/spack/repos/builtin/packages/icedtea/package.py:
```diff
@@ -0,0 +1,163 @@
+    depends_on('pkg-config', type='build')
+    depends_on('gmake', type='build')
+    depends_on('cups')
+    #X11 deps required for building even when headless
```
Flake8 will want a space after that comment symbol.

adamjstewart commented on var/spack/repos/builtin/packages/icedtea/package.py:
```diff
@@ -0,0 +1,163 @@
+    depends_on('gmake', type='build')
+    depends_on('cups')
+    #X11 deps required for building even when headless
+    depends_on('libx11', when='~X', type='build')
```
Wait, it needs `libx11` for both `+X` and `~X`? That doesn't make sense.

snehring replied:
Yeah it's weird, it checks for X deps even if --enable-headless is specified.

adamjstewart replied:
That's frustrating. Well I'm glad you added the dependencies in that case.

adamjstewart commented on var/spack/repos/builtin/packages/icedtea/package.py:
```diff
@@ -0,0 +1,163 @@
+    depends_on('jpeg')
+    depends_on('lcms')
+    depends_on('zlib')
+    depends_on('alsa-lib')
```
I think you're missing a `fastjar` dependency. @hartzell couldn't get it to build without it. You can steal mine from #2624.

snehring replied:
the building openjdk/icedtea should have jar included. I can add fastjar anyway if you'd like.

adamjstewart replied:
So you're saying that in order to build `icedtea`, you need a pre-existing working Java installation, and that should already come with `jar` or `fastjar`? That makes sense. Honestly, the only reason we started working on `icedtea` in #2624 was because `jdk` has weird licensing and we wanted a way to bootstrap `openjdk` from source, but it turns out `icedtea` needs to be bootstrapped too.

snehring replied:
I considered creating a binary icedtea/openjdk (sort of lilke gentoo does with openjdk-bin I think) with the sole purpose of being used to bootstrap this, do you think that's a good idea?

adamjstewart replied:
Hmm, I'll let @tgamblin comment on that.

snehring replied:
On a lark I decided to try to have this build its openjdk bootstrap env with the oracle jdk, and that actually appears to be working. I'm going to test a bit more, but if this works consistently would you be okay with an oracle jdk dependency?

adamjstewart replied:
I'm wondering how that will affect virtual dependencies. Anyway, it won't solve the licensing problem. But it seems to be the right thing to do.

adamjstewart commented on var/spack/repos/builtin/packages/icedtea/package.py:
```diff
@@ -0,0 +1,163 @@
+    def configure_args(self):
+        os.environ['POTENTIAL_CXX'] = os.environ['CXX']
+        os.environ['POTENTIAL_CC'] = os.environ['CC']
+        os.environ['WGET'] = self.spec['wget'].prefix+"/bin/wget"
```
You can now do `self.spec['wget'].command.path`.

adamjstewart commented on var/spack/repos/builtin/packages/icedtea/package.py:
```diff
@@ -0,0 +1,163 @@
+        args.append('--disable-bootstrap')
+        args.append('--disable-system-pcsc')
+        args.append('--disable-system-sctp')
+        args.append('--disable-system-kerberos')
```
Instead of appending 20 times, it might be simpler to extend once.

adamjstewart commented:
Another thing that needs to be done. We should add `provides('java')` to `jdk`, set a default `java` in `etc/spack/defaults/packages.yaml` (should probably go with `jdk` until we can get `openjdk` or `icedtea` to bootstrap from source), and replace all `depends_on('jdk')` with `depends_on('java')`.

snehring commented:
Switched to building the bootstrap jdk with the existing jdk package, so now at least this can build with whats included in spack. Also replaced all `depends_on('jdk')` with `depends_on('java')`. Also replaced any entries of `self.spec['jdk']` with `self.spec['java']` hopefully that 'just works' as expected.

snehring commented:
I am fairly certain I've gotten rid of every >79 char line that isn't a checksum now.

snehring commented:
Was there anything else on this you'd like me to do?

adamjstewart commented on var/spack/repos/builtin/packages/bazel/package.py:
```diff
@@ -42,7 +42,7 @@ class Bazel(Package):
     version('0.2.2', '644bc4ea7f429d835e74f255dc1054e6')
 
-    depends_on('jdk@8:')
+    depends_on('java@8')
```
It won't work with any newer version of java?

snehring replied:
whoops. That's a typo for sure.

adamjstewart commented on var/spack/repos/builtin/packages/jdk/package.py:
```diff
@@ -58,6 +58,9 @@ class Jdk(Package):
     version('7u80-b0', '6152f8a7561acf795ca4701daa10a965')
 
+    provides('java@8', when='@8u66-b17' '@8u73-b02' '@8u92-b14' '@8u131-b11')
+    provides('java@7', when='@7u80-b0')
```
I don't think this works? Why not say:
```python
provides('java@8', when='@8u0:8u999')
provides('java@7', when='@7u0:7u999')
```
You can confirm that this works like so:
```
$ spack python
Spack version 0.10.0
Python 2.7.13, Darwin x86_64
>>> from spack.version import *
>>> Version('8u66-b17') > Version('8u0')
True
>>> Version('8u66-b17') < Version('8u999')
True
```

snehring replied:
I wasn't sure how the versions would get parsed so I just made them explicit. I will change to the range.

adamjstewart commented on var/spack/repos/builtin/packages/icedtea/package.py:
```diff
@@ -0,0 +1,163 @@
+        run_env.set('JAVA_HOME', self.spec.prefix)
+
+    def setup_dependent_environment(self, spack_env, run_env, dependent_spec):
+        spack_env.set('JAVA_HOME', self.spec.prefix)
```
For the record, `self.spec.prefix == self.prefix` if you want to make it shorter.

adamjstewart commented:
If you check the failing Travis check, you'll see:
```
var/spack/repos/builtin/packages/icedtea/package.py:88: [E501] line too long (151 > 79 characters)
var/spack/repos/builtin/packages/icedtea/package.py:92: [E501] line too long (151 > 79 characters)
var/spack/repos/builtin/packages/icedtea/package.py:96: [E501] line too long (151 > 79 characters)
var/spack/repos/builtin/packages/icedtea/package.py:100: [E501] line too long (151 > 79 characters)
var/spack/repos/builtin/packages/icedtea/package.py:104: [E501] line too long (151 > 79 characters)
var/spack/repos/builtin/packages/icedtea/package.py:108: [E501] line too long (151 > 79 characters)
var/spack/repos/builtin/packages/icedtea/package.py:112: [E501] line too long (151 > 79 characters)
var/spack/repos/builtin/packages/icedtea/package.py:116: [E501] line too long (151 > 79 characters)
var/spack/repos/builtin/packages/icedtea/package.py:120: [E501] line too long (151 > 79 characters)
```
You can check this locally by installing flake8 and running `spack flake8`.

snehring commented:
For some reason I thought checksums were exempt from the line length restriction. I've corrected this.

adamjstewart commented:
> For some reason I thought checksums were exempt from the line length restriction. I've corrected this.

They are, but only if they occur on a single line. See [here](https://github.com/LLNL/spack/blob/develop/lib/spack/spack/cmd/flake8.py#L54) for a list of exemptions. Any line that starts with `version` or `resource` is exempt, but if you put the checksum on a different line, it is no longer exempt. Lines that look like URLs are also exempt.
